core/src/test/java/org/apache/calcite/test/SqlFunctionsTest.java

```diff
@@ -810,6 +826,674 @@ private void thereAndBack(byte[] bytes) {
+                SqlFunctions.PathContext
+                    .withReturned(SqlFunctions.PathMode.STRICT, Collections.singletonList("bar")),
+                SqlJsonQueryWrapperBehavior.WITH_CONDITIONAL_ARRAY,
+                SqlJsonQueryEmptyOrErrorBehavior.NULL,
+                SqlJsonQueryEmptyOrErrorBehavior.NULL),
+        is("[\"bar\"]")
+    );
+  }
+
+  @Test public void testJsonize() {
+    assertThat(
+        SqlFunctions.jsonize(new HashMap<>()),
+        is("{}")
+    );
+  }
+
+  @Test public void testDejsonize() {
+    assertThat(
+        SqlFunctions.dejsonize("{}"),
+        is(Collections.emptyMap())
+    );
+    assertThat(
+        SqlFunctions.dejsonize("[]"),
+        is(Collections.emptyList())
+    );
+
+    // expect exception thrown
+    try {
+      SqlFunctions.dejsonize("[}");
+      fail("expect exception, but not");
+    } catch (Exception ignored) {
+      // ignored
+    }
+  }
+
+  @Test public void testJsonObject() {
+    assertThat(
+        SqlFunctions.jsonObject(SqlJsonConstructorNullClause.NULL_ON_NULL),
+        is("{}")
+    );
+    assertThat(
+        SqlFunctions.jsonObject(SqlJsonConstructorNullClause.NULL_ON_NULL,
+            "foo",
+            "bar"),
+        is("{\"foo\":\"bar\"}")
+    );
+    assertThat(
+        SqlFunctions.jsonObject(SqlJsonConstructorNullClause.NULL_ON_NULL,
+            "foo",
+            null),
+        is("{\"foo\":null}")
+    );
+    assertThat(
+        SqlFunctions.jsonObject(SqlJsonConstructorNullClause.ABSENT_ON_NULL,
+            "foo",
+            null),
+        is("{}")
+    );
+  }
+
+  @Test public void testJsonObjectAggAdd() {
+    Map<String, Object> map = new HashMap<>();
+    Map<String, Object> expected = new HashMap<>();
+    SqlFunctions.jsonObjectAggAdd(map,
+        "foo", "bar", SqlJsonConstructorNullClause.NULL_ON_NULL);
+    expected.put("foo", "bar");
+    assertThat(map,
+        is(expected)
+    );
+    SqlFunctions.jsonObjectAggAdd(map,
+        "foo1", null, SqlJsonConstructorNullClause.NULL_ON_NULL);
+    expected.put("foo1", null);
+    assertThat(map,
+        is(expected)
+    );
+    SqlFunctions.jsonObjectAggAdd(map,
+        "foo2", null, SqlJsonConstructorNullClause.ABSENT_ON_NULL);
+    assertThat(map,
+        is(expected)
+    );
+  }
+
+  @Test public void testJsonArray() {
+    assertThat(
+        SqlFunctions.jsonArray(SqlJsonConstructorNullClause.NULL_ON_NULL),
+        is("[]")
+    );
+    assertThat(
+        SqlFunctions.jsonArray(SqlJsonConstructorNullClause.NULL_ON_NULL,
+            "foo"),
+        is("[\"foo\"]")
+    );
+    assertThat(
+        SqlFunctions.jsonArray(SqlJsonConstructorNullClause.NULL_ON_NULL,
+            "foo",
+            null),
+        is("[\"foo\",null]")
+    );
+    assertThat(
+        SqlFunctions.jsonArray(SqlJsonConstructorNullClause.ABSENT_ON_NULL,
+            "foo",
+            null),
+        is("[\"foo\"]")
+    );
+  }
+
+  @Test public void testJsonArrayAggAdd() {
```

Comment:
Would you please ensure the tests print human-understandable messages in case of failures?